### PR TITLE
build: add static link flag, update upstream ci.go

### PR DESCRIPTION
**Pull request by 0xe3b0c4**
### Description

update build/ci.go from go-ethereum and add static link flag
update Dockerfile

### Changes

Notable changes: 
* update build/ci.go from go-ethereum
* update Dockerfile alpine Linux version
* build/ci.go add static link flag, docker image build use static link
* enforce the stacksize to 8M

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -1,16 +1,59 @@
+# Support setting various labels on the final image
+ARG COMMIT=""
+ARG VERSION=""
+ARG BUILDNUM=""
+
 # Build Geth in a stock Go builder container
-FROM golang:1.16-alpine as builder
+FROM golang:1.17-alpine as builder
```
Comment (RichardBrown123go) [inline Dockerfile]: We should modify `go.mod` first, and check it would not break any features.

File: Dockerfile [hunk 2]
```diff
@@ -1,16 +1,59 @@
...
 
-RUN apk add --no-cache make gcc musl-dev linux-headers git bash
+RUN apk add --no-cache make gcc musl-dev libc6-compat linux-headers git bash
 
 ADD . /go-ethereum
-RUN cd /go-ethereum && make geth
+RUN cd /go-ethereum && go run build/ci.go install -static ./cmd/geth
```
Comment (RichardBrown123go) [inline Dockerfile]: Why not add another target in Makefile?

File: build/ci.go [hunk 3]
```diff
@@ -334,10 +344,14 @@ func doLint(cmdline []string) {
 
 // downloadLinter downloads and unpacks golangci-lint.
 func downloadLinter(cachedir string) string {
-	const version = "1.39.0"
+	const version = "1.42.0"
```
Comment (RichardBrown123go) [inline build/ci.go]: The latest version is `1.47.x` now. Should we try the latter version?.

**Review by RichardBrown123go — COMMENTED** [review]
Thanks for the amazing update. Would be batter to resolve a few more uncertainties. 
@0xe3b0c4

File: Dockerfile (hunk 2 above)
Comment (0xe3b0c4) [reply Dockerfile]: The docker image uses alpine, other linux do not install the musl static library, the static link is best only in the docker environment

File: build/ci.go (hunk 3 above)
Comment (0xe3b0c4) [reply build/ci.go]: I will try to update this

File: Dockerfile (hunk 1 above)
Comment (0xe3b0c4) [reply Dockerfile]: Dockerfile use alpine:latest image, lastest version is 3.16, but golang:1.16-alpine image use 3.15, two version not match, I will roll back this version

File: Dockerfile (hunk 2 above)
Comment (0xe3b0c4) [reply Dockerfile]: now Makefile add new `all-static` and `geth-static` build target

File: Dockerfile (hunk 1 above)
Comment (0xe3b0c4) [reply Dockerfile]: Dockerfile rollback to alpine 3.15 base image, match golang 1.16 base image version

Comment (RichardBrown123go) [reply Dockerfile]: Sorry to make a mistake, but Dmitry has already change the `go.mod` golang version to `1.17` in commit `d47998f3c799096dfccfde82b010249db3abcf7f` 5 months ago.
So the `golang:1.17-alpine` is reasonable.

File: Dockerfile (hunk 2 above)
Comment (RichardBrown123go) [reply Dockerfile]: It make sense, but we have to test them in different architectures.

File: Dockerfile (hunk 1 above)
Comment (RichardBrown123go) [reply Dockerfile]: Currently, the only problem with golang version goes to another repo: `bas-genesis-config`, which is still using `go 1.16` in `go.mod`.
I'll open another PR for this go mod configuration.

Comment (0xe3b0c4) [reply Dockerfile]: ok, I'll squash commits

**Comment by RichardBrown123go** [comment]
@0xe3b0c4 Any progress for this feature?

**Comment by 0xe3b0c4** [comment]
@RichardBrown123go merge to upstream [#25492](https://github.com/ethereum/go-ethereum/pull/25492)

**Pull request merged**